compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/SharedVariablesLowering.kt

```diff
@@ -80,7 +80,8 @@ class SharedVariablesLowering(val context: BackendContext) : FunctionLoweringPas
                 override fun visitVariable(declaration: IrVariable, data: IrDeclarationParent?) {
                     declaration.acceptChildren(this, data)
 
-                    if (declaration.isVar) {
+                    // `val` locals can be initialized in a lambda inlined into `run` or similar.
+                    if (declaration.isVar || declaration.initializer == null) {
```

Comment:
What about such cases:
```
 val z: Int
    z = 1
```
seems we also should check kind of IrValueAccessExpression